Market analysis

datw0258 opened the pull request:
I uploaded the data to the box folder. I was having trouble getting the pathing correct to access it from the functions, so I set a specific path. Let me know if you have any suggestions for this to work generally.

datw0258 committed:
Merge pull request #2 from arfc/master

syncing from original

datw0258 committed:
creating market folder

datw0258 committed:
uploading education market analysis

datw0258 committed:
delete test file

datw0258 committed:
create __init__.py

datw0258 committed:
add __init__.py

datw0258 committed:
add __init__.py

datw0258 committed:
add __init__.py

pep8speaks commented:
Hello @datw0258! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2020-07-07 17:37:40 UTC

datw0258 commented:
Hm, the PEP8 website I used for these did not catch these. Making changes now

datw0258 committed:
pep8 fixes

datw0258 committed:
pep8 functions_2012.py

datw0258 committed:
pep8 functions_2016.py

datw0258 committed:
pep8 functions_2018.py

datw0258 committed:
pep8 init fix

datw0258 committed:
pep8 init

datw0258 commented:
The __init__.py PEP8 issues are giving me some trouble. I make a commit to fix them, and the file does not change

samgdotson reviewed (changes requested):
The notebooks are clean and it's great that you've created functions to do the analysis in separate files -- and written tests!

This PR could be much smaller, though. It looks to me like the functions in your ``functions_[year].py`` files are all basically the same with some slightly different hard coded values, same for your notebooks. 
This is breaking the "D.R.Y" principle -- Don't Repeat Yourself.

It would be great if you could condense everything into a single notebook, functions file, and tests.

edit: Regarding PEP8, I'm not sure what website you're using to make PEP8 corrections. It's complaining about having two blank lines when there should only be one. I recommend doing:

```bash
pip install autopep8
autopep8 --in-place --aggressive --aggressive __init__.py
```
This fixed it for me right away.

Finally, since you're going to put everything in one notebook -- just call the notebook something like ``education_energy_markets.ipynb``. Put your functions file in the same directory call it ``eia.py`` and put your tests in the ``tests`` folder that already exists.

datw0258 committed:
Delete test_functions_2008.py

datw0258 committed:
Delete __init__.py

datw0258 committed:
Delete functions_2008.py

datw0258 committed:
Delete eia_results_2008.ipynb

datw0258 committed:
Delete functions_2012.py

datw0258 committed:
Delete eia_results_2012.ipynb

datw0258 committed:
Delete test_functions_2012.py

datw0258 committed:
Delete __init__.py

datw0258 committed:
Delete functions_2016.py

datw0258 committed:
Delete eia_results_2016.ipynb

datw0258 committed:
Delete test_functions_2016.py

datw0258 committed:
Delete __init__.py

datw0258 committed:
Add files via upload

datw0258 committed:
Delete functions_2018.py

datw0258 committed:
Delete test_functions_2018.py

datw0258 committed:
Delete __init__.py

datw0258 committed:
Delete eia_results_2018.ipynb

datw0258 committed:
pep8 fixes

datw0258 committed:
Create __init__.py

datw0258 committed:
Add files via upload

datw0258 committed:
update test_functions.py

datw0258 committed:
pep8 fixes eia.py

datw0258 committed:
final pep8 fix eia.py

samgdotson commented on data_processing/market/eia.py:
```diff
@@ -0,0 +1,2087 @@
+            f"Not a valid year. Possibilities include: {*allowed_years,}."
+        )
+
+    sheet = path + 'eia_generation_' + year + '.csv'
+
+    Generation = sheet
+
+    with open(Generation, 'r') as i:
+
+        generation = i.read().splitlines()
```
```suggestion
   Generation = path + 'eia_generation_' + year + '.csv'

    with open(Generation, 'r') as i:

        generation = i.read().splitlines()
```

samgdotson replied:
This can be done in ``pandas``.

samgdotson commented on data_processing/market/eia.py:
```diff
@@ -0,0 +1,2087 @@
+                                    {'Total Generation': new_total}
+                                )
+
+                    elif (name_old, region_old) not in names_checked:
+
+                        name_old = name_old + ' ' + region_old
+
+                        counter = 0
+
+                        for fuel in name_dict:
+
+                            for prime in name_dict[fuel]:
+
+                                counter += name_dict[fuel][prime]
+
+                        name_dict.update({'Total Generation': counter})
+
+                        global_dict.update({name_old: name_dict})
+
+                name_dict = dict()
+
+                if fuel_type in name_dict:
+
+                    if mover in name_dict[fuel_type]:
+
+                        name_dict[fuel_type][mover] += amount
+
+                    elif mover not in name_dict[fuel_type]:
+
+                        name_dict[fuel_type].update({mover: amount})
+
+                elif fuel_type not in name_dict:
+
+                    name_dict.update({fuel_type: {mover: amount}})
+
+            name_old = name
+
+            region_old = region
+
+    del global_dict['']
```
@katyhuff  posted in the pro-tips slack channel about leaving "generous amounts of comments in your code." 

This is a great example as to why. 

Also, the way I understand it, you're manipulating a ``.csv`` file and turning it into a "nested dictionary." I think this can and should be done using the ``pandas`` library. If it's a higher dimensional dataframe, then I suggest using ``XArray``. 

The benefit of using these powerful libraries is that it makes your code much more comprehensible. If we recall the Zen of Python: Namespaces are a honking great idea.

katyhuff replied:
I like comments where docstrings aren't possible. But, I prefer documentation that can be tested (that is, clear variable names and function structure that is self-explanatory) over comments. (Robert C. Martin style). I do agree that you need to think of your future self (and comments can contribute) but given a choice, I prefer clear variable names, extremely small well-named functions, and elaborate docstrings.

samgdotson commented on data_processing/market/eia.py:
```diff
@@ -0,0 +1,2087 @@
+            'https://www.hydrogen.energy.gov/pdfs/review16/tv016_saur_2016' +
+            '_p.pdf'
+        )
+
```
I think you're missing a ``return`` statement.

samgdotson reviewed (changes requested):
First, thank you for making the changes I requested last time. Having a single notebook makes it much cleaner. You clearly put a lot of effort into writing these functions and that's awesome. There are still some things that need to change before I can merge the PR. The comments below are detailed but I think they will be helpful. 

#### Tests
1. Your PR breaks the tests that I wrote previously. Currently, in the main branch if I do ``pytest ./data_processing/`` everything passes. This is no longer the case in your PR. 
2. Not all of your tests pass. Fortunately the ones that fail are the ones where you read from a file. This is easily remedied by adding a test data file in the ``tests`` folder (I know this seems counter to what I said before about not having data in the repo). 
3. Fixing your tests alone will not fix the tests that already exist which leads me to....

#### Organization

I should have been clearer about what I wanted in the previous review. The repo is currently set up like this:
``pride``
-- ``data_processing``
-- -- ``notebooks.ipynb``
-- -- ``tests``
-- -- -- ``__init__.py``
-- ``other_folders``

I want to keep this organization roughly the same. By virtue of condensing your analysis to a single notebook it can logically stay in the ``data_processing`` folder *without the need for a ``markets`` subfolder*. The tests your wrote can then go directly into the ``tests`` folder that already exists in ``data_processing``. So the ideal organization will look like this:

``pride``
-- ``data_processing``
-- -- ``notebooks.ipynb``
-- -- ``eia_notebook.ipynb``
-- -- ``tests``
-- -- -- ``__init__.py``
-- -- -- ``dotson_tests.py``
-- -- -- ``atwater_tests.py`` (don't call it this, purely illustrative)
-- ``other_folders``

The existence of two ``tests`` folders confuses ``pytest``. 

#### Code 

1. Comments 
I mention below that @katyhuff recommends commenting your code. This benefits others reading it, but it also helps you! Years from now when you've forgotten why you made the choices you did you will save yourself a lot of time by commenting your code. E.g. 

```python
#  This for loop does this repeated calculation
for x in range(10):
   calculate()
```
2. Pandas?
Pandas is a super powerful library for manipulating ``csv`` files. It's totally possible that ``pandas`` can't do what you want it to, but it's highly unlikely. With just a few lines of code I can open a file, turn it into a dataframe and start messing with it.
This 

```python
import pandas as pd
dataframe = pd.read_csv("path-to-file")
```
is superior to 
```python
with open('path-to-file', r) as file:
    lines = file.readlines().split()

datapoints = lines.split(',')
...
```
You can do all kinds of magic with ``pandas`` like interpolation, data visualization, and much more. 
If you need some guidance, I use ``pandas`` in practically all of my notebooks. 

[10-minute Tutorial](https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html)

datw0258 commented:
I will start by committing changes for readability (commenting code etc.), then I will learn pandas and see if I can manipulate it to fit my code. Thank you for these suggestions!

katyhuff commented:
To be clear, I actually don't prefer inline code comments. (yes, I suppose I prefer them over nothing, but it's not my way to litter the code with comments.) I follow the Robert C. Martin Clean Code approach.

- small functions (yes, this means breaking big functions up in to lots of little ones)
- clear variable names
- very detailed, standardized, function **docstrings** with all appropriate external citations as well as default values and clear names for all parameters.

samgdotson commented:
Then I'd suggest starting with ``pandas``, @datw0258. Using that library will satisfy the unit-functions and make your code less opaque without the need for inline comments.

datw0258 committed:
Rename data_processing/market/eia.py to data_processing/eia.py

datw0258 committed:
Delete test_functions.py

datw0258 committed:
Delete __init__.py

datw0258 committed:
Delete education_market_analysis.ipynb

datw0258 committed:
Update eia.py

datw0258 committed:
moving test_eia_functions.py

datw0258 committed:
updating eia.py to allow tests to pass

not sure if this will translate out of windows

datw0258 committed:
adding a sample data csv

datw0258 committed:
moving eia_market_analysis.ipynb and adding years

datw0258 committed:
Merge pull request #3 from datw0258/market-analysis-workspace

Market analysis workspace

datw0258 committed:
pep8

datw0258 commented:
I changed my spreadsheet work into pandas (thank you for the tip). I also changed the organization to your request and added several years to my analysis.

As far as the tests are concerned, let me know if there are still problems. I created a sample version of one of the `.csv` files, but I am not sure the code I wrote to access it translates off the windows  operating platform.

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,2999 @@
+
+# path = 'C:\\Users\\Atwater\\research\\data\\eia_generation_'
+
+path = '.\\tests\\eia_sample_'
```
```suggestion
path = './tests/eia_sample_'
```

The way the path is written right now works on windows, but not on Linux. This suggestion fixes the issue.

samgdotson replied:
Actually, making this a global variable breaks the functions because the only file it knows about is the one in ``tests``. 

Instead, tell people where they should download the data and set path

``path = "../../data/``

Up one directory to ``pride``, then up one more to ``research`` then go to ``data``.

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,2999 @@
+        '2018': 'Net Generation (Megawatthours)'
+    }
+
+    generation = pd.read_csv((path + year + '.csv'),
+                             header=headers[year],
+                             usecols=['Operator Name',
+                                      'NERC Region',
+                                      'NAICS Code',
+                                      'Reported Prime Mover',
+                                      'Reported Fuel Type Code',
+                                      str(netGen[year])])
```
This does not work in Linux because the line endings in Windows don't translate. 

The column names are actually "Reported\n Prime Mover"

Where does the "\n" go? It's inconsistent and depends on the column. Some have it, some don't. I'd recommend determining the column numbers and using those instead, it's much more consistent.

datw0258 replied:
I forgot to mention I updated the `csv` files in box to have no line breaks. This should be fixed by those files. Apologies.

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,2999 @@
+                                      'Reported Fuel Type Code',
+                                      str(netGen[year])])
+
+    generation = generation.loc[generation['NAICS Code'] == 611]
+
+    generation['University'] = generation['Operator Name'].astype(
+        str) + ' ' + generation['NERC Region']
+
+    generation = generation.drop(
+        ['Operator Name', 'NERC Region'],
+        axis=1
+    )
```
You're essentially renaming a column by creating a new column and dropping two others. I couldn't find where "NERC Region" came up in your analysis (it wasn't broken up by region anywhere) so we can probably leave it alone.

This is done easily in pandas with the following
```suggestion
    generation = generation.loc[generation['NAICS Code'] == 611]

    generation.rename(columns={'Operator Name':'University'}, inplace=True)

```

samgdotson replied:
Nice use of masking to select the "NCAIS code" :smile:

datw0258 replied:
So I decided to include the NERC region because there were a few instances I noticed where the Operator is the same University name, but its actually a different institution. For instance "University of Illinois" is listed twice in all of the EIA 293s, but upon further inspection one of them is the University of Illinois Chicago and one is Urbana-Champaign. The only distinguishing column was the NERC region. If I removed the region it combined the numbers and University of Illinois became the largest producer.
Do you have any better solutions to combat this problem besides hard-coding all the duplicates?

samgdotson replied:
I made the suggestion because it wasn't obvious to me why the NERC region
was necessary. Based on your description of the problem it is necessary to
distinguish institutions, so I think this solution is fine. I might do:

```python
universities = generation['NAICS Code'] == 611
generation = generation.loc[universities]
generation['University'] = generation['Operator Name'] + ' '  + generation['NERC Region']
generation.drop(['Operator Name', 'NERC Region'], axis=1, inplace=True)
```
Since they're already strings you can just use the ``+`` operator. I added
a line for masking (the first line) because it's clearer
to the reader why you're selecting that code. (At least, I'm assuming
they're universities).

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,2999 @@
+        '2018': 5
+    }
+
+    netGen = {
+        '2008': 'NET GENERATION (megawatthours)',
+        '2009': 'NET GENERATION (megawatthours)',
+        '2010': 'NET GENERATION (megawatthours)',
+        '2011': 'Net Generation (Megawatthours)',
+        '2012': 'Net Generation (Megawatthours)',
+        '2013': 'Net Generation (Megawatthours)',
+        '2014': 'Net Generation (Megawatthours)',
+        '2015': 'Net Generation (Megawatthours)',
+        '2016': 'Net Generation (Megawatthours)',
+        '2017': 'Net Generation (Megawatthours)',
+        '2018': 'Net Generation (Megawatthours)'
+    }
```
```suggestion
```
This is redundant. If every file has the same column name you can just put it 
in ``usecols``

samgdotson replied:
```suggestion
if int(year) < 2011:
    netGen = 'NET GENERATION (megawatthours)'
elif int(year) >= 2011: 
    netGen = 'Net Generation (Megawatthours)'
```

I see that the column names changed afte 2010... unfortunate. Since there are only two options I'd do this.

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,2999 @@
+                                      'NAICS Code',
+                                      'Reported Prime Mover',
+                                      'Reported Fuel Type Code',
+                                      str(netGen[year])])
```
```suggestion
                                    netGen)
```

Making it consistent with previous suggestion

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,2999 @@
+
+        iterable_dataframe = generation.loc[
+            generation['University'] == uni
+        ]
+
+        fuel_type_list = iterable_dataframe[
+            'Reported Fuel Type Code'
+        ].unique().tolist()
+
+        for fuel in fuel_type_list:
+
+            mover_dict = dict()
+
+            fuel_dataframe = iterable_dataframe.loc[
+                generation['University'] == uni
+            ].loc[
+                iterable_dataframe['Reported Fuel Type Code'] == fuel
+            ]
+
+            mover_list = fuel_dataframe[
+                'Reported Prime Mover'
+            ].values.tolist()
+
+            netgen_list = fuel_dataframe[
+                netGen[year]
+            ].values.tolist()
+
+            for i in range(len(mover_list)):
+
+                totalgen += netgen_list[i]
+
+                mover_dict.update({mover_list[i]: netgen_list[i]})
+
+            fuel_dict.update({fuel: mover_dict})
+
+        fuel_dict.update({'Total Generation': totalgen})
+
+        final.update({uni: fuel_dict})
+
+    final
```
This can be done much more succinctly with Pandas. 

Not an exact suggestion because there are multiple Pandas solutions, I would start by using the ``groupby`` function. 

```python

#  get groups of items that all have the same naics code
naics = generation.groupby('NAICS Code')

#  get the universities
universities = naics.get_group(611)

#  regroup again by either fuel or something else
```

Here is a [good blog post](https://www.shanelynn.ie/summarising-aggregation-and-grouping-data-in-python-pandas/) that shows what I mean.

samgdotson reviewed (changes requested):
Good news! The PR improves with every commit. 

I'm still hesitant to approve it for a couple of reasons. There are some issues with the ``path``. It should be able to find the data folders. If you read the comments I'm a little wishy-washy, my recommendation is this: make the path a function argument so that users can point to where their data lives. 

For the most part your code is "right." It's tested and seems stable. Clearly it works because you have some great plots in the notebook. It's also hard to read. I tried to find a few places where Pandas would dramatically reduce the line count. I really think you should try and find a way to incorporate Pandas into your code. 

The [Zen of Python](https://www.python.org/dev/peps/pep-0020/) is oft quoted. "Simple is better than complex."

This might be a personal thing for me, but I like closing functions with ``return`` even if it returns nothing. I've gotten lost a couple times because I couldn't tell when a function ended (nested functions do exist in some cases). 

Thanks for making so many changes!

katyhuff commented:
Just wanted to pop in here to say (1) this PR is starting to look really polished and (2) I agree with @samgdotson about return statements. It's so important that all functions return something, actually, that (in Python) if there are no return statements, then the function returns something anyway: None.

samgdotson commented:
I made the suggestion because it wasn't obvious to me why the NERC region
was necessary. Based on your description of the problem it is necessary to
distinguish institutions, so I think this solution is fine. I might do:

```python
universities = generation['NAICS Code'] == 611
generation = generation.loc[universities]
generation['University'] = generation['Operator Name'] + ' '  +
generation['NERC Region']
generation.drop(['Operator Name', 'NERC Region'], axis=1, inplace=True)
```
Since they're already strings you can just use the ``+`` operator. I added
a line for masking (the first line) because it's clearer
to the reader why you're selecting that code. (At least, I'm assuming
they're universities).

On Mon, Jul 6, 2020 at 10:05 AM datw0258 <notifications@github.com> wrote:

> *@datw0258* commented on this pull request.
> ------------------------------
>
> In data_processing/eia.py
> <https://github.com/arfc/pride/pull/68#discussion_r450286410>:
>
> > +    generation = generation.loc[generation['NAICS Code'] == 611]
> +
> +    generation['University'] = generation['Operator Name'].astype(
> +        str) + ' ' + generation['NERC Region']
> +
> +    generation = generation.drop(
> +        ['Operator Name', 'NERC Region'],
> +        axis=1
> +    )
>
> So I decided to include the NERC region because there were a few instances
> I noticed where the Operator is the same University name, but its actually
> a different institution. For instance "University of Illinois" is listed
> twice in all of the EIA 293s, but upon further inspection one of them is
> the University of Illinois Chicago and one is Urbana-Champaign. The only
> distinguishing column was the NERC region. If I removed the region it
> combined the numbers and University of Illinois became the largest producer.
> Do you have any better solutions to combat this problem besides
> hard-coding all the duplicates?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/arfc/pride/pull/68#discussion_r450286410>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKSJ4WO2NQSGU3G4632WHILR2HR4ZANCNFSM4OJN54FA>
> .
>


-- 
Samuel G. Dotson
Master of Science in Nuclear Engineering 2021
Bachelor of Science in Engineering Physics
University of Illinois at Urbana-Champaign
Office: 133 Computer Applications Building, 605 E Springfield Ave
Pronouns: he/him/his
sgd2@illinois.edu

datw0258 committed:
updating eia.py as per review

datw0258 commented:
I made the updates to most of your reviews (I am still working on succinctly doing `generation`). I am unclear if this will fix the errors running `tests` though.

datw0258 committed:
bug fixes and reducing eia.py

datw0258 committed:
updating education_market_analysis.ipynb

datw0258 commented:
I removed almost all of the hard-coded bits using pandas to get the labels easier. This significantly reduced the line count (thank you for the tip).

*Grammar Edit*

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,1526 @@
+
+    """
+
+    data = generation('2018')
```
```suggestion
    data = generation(year=year, path=path, prefix=prefix)
```

Does this function need to be for 2018 only? It might be interesting to see how the capacity evolves over time.

datw0258 replied:
I did it for 2018 because I could not find accurate data for capacity factors for the earlier years (and I know they are evolving by year).

samgdotson replied:
Got it, we can still set the default value to "2018" rather than hard coding it.

datw0258 replied:
In my `plot_data` function for the capacity argument, I have it raise an error if 2018 is not selected as the year. Should I remove this too in case someone wanted to make approximations based on current capacity factors?

samgdotson replied:
Yeah, that'd be good. I couldn't catch every instance, but where you can make a function generic,  you should.

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,1526 @@
+    return cf
+
+
+def university_capacity():
```
```suggestion
def university_capacity(year: str, path=path, prefix=fileprefix):
```

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,1526 @@
+    return fuel_breakdown
+
+
+def university_top_renewables(year: str):
```
```suggestion
def university_top_renewables(year: str, path=path, prefix=fileprefix):
```

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,1526 @@
+        netGen = 'Net Generation (Megawatthours)'
+
+        headers = 5
+
+    generation = pd.read_csv((path + year + '.csv'),
```
```suggestion

    filename = fileprefix + year + ".csv"
    generation = pd.read_csv((path + filename),
```
This is better because it allows people to point to different folders where their data may be stored.

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,1526 @@
+    return mover
+
+
+def generation(year: str):
```
```suggestion
def generation(year: str, path=path, prefix=fileprefix):
```

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,1526 @@
+import matplotlib.pyplot as plt
+import pandas as pd
+
+path = '../../data/eia_generation_'
```
```suggestion
path = "../../data/"
fileprefix = "eia_generation_"
```
We don't want the path to include a file name because it's not really a path. I.e. I can't point to a different folder with the datafiles.

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,1526 @@
+    return final
+
+
+def university_top_producers(year: str):
```
```suggestion
def university_top_producers(year: str, path=path, prefix=fileprefix):
```

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,1526 @@
+            f"Not a valid year. Possibilities include: {*allowed_years,}."
+        )
+
+    inp = year
+
+    global_dict = generation(inp)
```
```suggestion

    global_dict = generation(year=year, path=path, prefix=prefix)
```

samgdotson commented on data_processing/eia.py:
```diff
@@ -0,0 +1,1526 @@
+
+    renewables_list = ['GEO', 'WND', 'SUN', 'WAT', 'OBG', 'OBL', 'OBS']
+
+    inp = year
+
+    global_dict = generation(inp)
```
```suggestion

    global_dict = generation(year=year, path=path, prefix=prefix)
```

samgdotson commented on data_processing/tests/test_eia_functions.py:
```diff
@@ -0,0 +1,539 @@
+from eia import split_dictionary
+from eia import plot_data
+from eia import plot_energy_change
+path = 'eia_sample_'
```
```suggestion
path = "./"
fileprefix= "eia_sample_"
```
The suggestions made in ``eia.py`` should be reflected here.

samgdotson commented on data_processing/tests/test_eia_functions.py:
```diff
@@ -0,0 +1,539 @@
+
+    """
+
+    top_list = list(university_top_renewables('2018'))
+
+    top_dict = university_top_renewables('2018')
```
```suggestion
    top_list = list(university_top_renewables('2018', path=path, prefix=fileprefix))

    top_dict = university_top_renewables('2018', path=path, prefix=fileprefix)
```

samgdotson commented on data_processing/tests/test_eia_functions.py:
```diff
@@ -0,0 +1,539 @@
+
+    """
+
+    top_list = list(university_top_producers('2018'))
+
+    top_dict = university_top_producers('2018')
```
```suggestion
    top_list = list(university_top_producers('2018', path=path, prefix=fileprefix))

    top_dict = university_top_producers('2018', path=path, prefix=fileprefix)
```

samgdotson commented on data_processing/tests/test_eia_functions.py:
```diff
@@ -0,0 +1,539 @@
+
+    """
+
+    top_list = list(university_capacity())
+
+    top_dict = university_capacity()
```
```suggestion
    top_list = list(university_capacity("2018", path=path, prefix=fileprefix))

    top_dict = university_capacity("2018", , path=path, prefix=fileprefix)
```

samgdotson reviewed (changes requested):
Your main code is *much* shorter, thank you! There are still three tests that do not run. All of them are due to a pathing issue. I've made some suggestions that I think will solve the problem for you by making the functions that 
1. Read a datafile or 
2. Call a function that reads a datafile

use parameters that specify the location and name of the files. The reason the tests don't run is because you made "path" a global variable in ``eia.py`` but ``tests_`` can't see that! The test data is in a different directory than the analyzed data. So, users can now change the "path" and "fileprefix" variables to reflect differences in their directory structure. It should also fix the tests (but I haven't checked). 

After the tests pass this PR is ready to merge.

datw0258 committed:
adding path and prefix to eia.py functions

datw0258 committed:
updating to reflect changes in eia.py

datw0258 commented:
I have updated the functions, now I will adjust the tests and then they should be good to go.

datw0258 committed:
updating eia_test_functions.py

datw0258 committed:
pep8 change for eia.py

datw0258 committed:
pep8 eia.py

samgdotson commented on data_processing/tests/test_eia_functions.py:
```diff
@@ -0,0 +1,614 @@
+from eia import plot_data
+from eia import plot_energy_change
+
+path_arg = './'
```
```suggestion
path_arg = './data_processing/tests/'
```

This fixes the tests, finally. The relative paths are a little confusing. The path to the sample data must be *relative to ``eia.py``* because that's where the function is actually called. Just "commit suggestion" and we'll be good to go!

samgdotson reviewed (changes requested):
You can just "commit suggestion" directly in the conversation to fix the tests, finally.

datw0258 committed:
Update data_processing/tests/test_eia_functions.py

Co-authored-by: Sam Dotson <44342873+samgdotson@users.noreply.github.com>

samgdotson commented on data_processing/tests/test_eia_functions.py:
```diff
@@ -13,7 +13,7 @@
 from eia import plot_energy_change
 
-path_arg = './'
+path_arg = './data_processing/tests'
```
```suggestion
path_arg = './data_processing/tests/'
```

samgdotson committed:
Update data_processing/tests/test_eia_functions.py

samgdotson commented:
Beautiful. 

Thanks for making all of these changes, @datw0258 ! The tests pass and the script is nicely cleaned up. Merging now.

samgdotson reviewed (approved):
lgtm merging
